onnx_models/conftest.py

```diff
@@ -0,0 +1,355 @@
+        logger.error("iree-import-onnx stdout:")
+        logger.error(ret.stdout.decode("utf-8"))
+        logger.error("iree-import-onnx stderr:")
+        logger.error(ret.stderr.decode("utf-8"))
+        raise IreeImportOnnxException(f"  '{onnx_path.name}' import failed")
+    return imported_mlir_path
+
+
+def convert_proto_elem_type_to_iree_dtype(etype):
+    if etype == onnx.TensorProto.FLOAT:
+        return "f32"
+    if etype == onnx.TensorProto.UINT8:
+        return "i8"
+    if etype == onnx.TensorProto.INT8:
+        return "i8"
+    if etype == onnx.TensorProto.UINT16:
+        return "i16"
+    if etype == onnx.TensorProto.INT16:
+        return "i16"
+    if etype == onnx.TensorProto.INT32:
+        return "i32"
+    if etype == onnx.TensorProto.INT64:
+        return "i64"
+    if etype == onnx.TensorProto.BOOL:
+        return "i1"
+    if etype == onnx.TensorProto.FLOAT16:
+        return "f16"
+    if etype == onnx.TensorProto.DOUBLE:
+        return "f64"
+    if etype == onnx.TensorProto.UINT32:
+        return "i32"
+    if etype == onnx.TensorProto.UINT64:
+        return "i64"
+    if etype == onnx.TensorProto.COMPLEX64:
+        return "complex<f32>"
+    if etype == onnx.TensorProto.COMPLEX128:
+        return "complex<f64>"
+    if etype == onnx.TensorProto.BFLOAT16:
+        return "bf16"
+    if etype == onnx.TensorProto.FLOAT8E4M3FN:
+        return "f8e4m3fn"
+    if etype == onnx.TensorProto.FLOAT8E4M3FNUZ:
+        return "f8e4m3fnuz"
+    if etype == onnx.TensorProto.FLOAT8E5M2:
+        return "f8e5m2"
+    if etype == onnx.TensorProto.FLOAT8E5M2FNUZ:
+        return "f8e5m2fnuz"
+    if etype == onnx.TensorProto.UINT4:
+        return "i4"
+    if etype == onnx.TensorProto.INT4:
+        return "i4"
+    return ""
+
+
+def convert_onnx_tensor_proto_to_numpy_dimensions(
+    tensor_proto: onnx.onnx_ml_pb2.TensorProto,
+) -> str:
+    # Note: turning dynamic dimensions into just 1 here, since we need
+    # a concrete (static) shape buffer of input data in the tests.
+    return tuple(
+        d.dim_value if d.HasField("dim_value") else 1 for d in tensor_proto.shape.dim
+    )
+
+
+def convert_onnx_tensor_proto_to_iree_type_string(
+    tensor_proto: onnx.onnx_ml_pb2.TensorProto,
+) -> str:
+    shape = tensor_proto.shape
+    # Note: turning dynamic dimensions into just "1" here, since we need
+    # a concrete (static) shape buffer of input data in the tests.
+    shape = "x".join(
+        [str(d.dim_value) if d.HasField("dim_value") else "1" for d in shape.dim]
+    )
+    dtype = convert_proto_elem_type_to_iree_dtype(tensor_proto.elem_type)
+    if shape == "":
+        return dtype
+    return f"{shape}x{dtype}"
+
+
+def get_onnx_model_metadata(onnx_path: Path):
+    # We can either
+    #   A) List all metadata explicitly
+    #   B) Get metadata on demand from the .onnx protobuf using 'onnx'
+    #   C) Get metadata on demand from the InferenceSession using 'onnxruntime'
+    # This is option B.
```

Comment:
Another option I'm considering is having test case import be an offline, tool-assisted step, instead of an online step that always runs as part of the tests. That way, we can have developers importing model tests decide explicitly what functions and signatures they want to test. It would be nice if there was only one obvious function and signature for each model that can be safely inferred though :)

Reply:
Made some functional updates, but I'll continue to iterate on the code style.

This is now using option C - getting the input and output signatures from `onnxruntime.InferenceSession`. The output signatures are now exactly what comes from onnxruntime, while the input signatures are still turning dynamic dims into 1.